[WKTR] Crash WKTR in setup if no screens returned from AppKit

zakariaridouh opened the pull request:
#### 06a683e38ea5a72ef50a454d65ef6779a90393ba
<pre>
[WKTR] Crash WKTR in setup if no screens returned from AppKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=291122">https://bugs.webkit.org/show_bug.cgi?id=291122</a>
<a href="https://rdar.apple.com/148636042">rdar://148636042</a>

Reviewed by David Kilzer and Alexey Proskuryakov.

We&apos;ve come across some situations where the `NSScreen screens` API
doesn&apos;t return any valid screens for us to use in WKTR (for example
post WindowServer crash, or sandboxing where the app can&apos;t connect to WS),
and we should not support this configuration.

Add a release assert with message to crash the process.

* Tools/WebKitTestRunner/mac/PlatformWebViewMac.mm:
(WTR::PlatformWebView::PlatformWebView):

Canonical link: <a href="https://commits.webkit.org/293315@main">https://commits.webkit.org/293315@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d56b47bdf534e0e6d18e74af267ed7098eb6b31

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98530 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18163 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8395 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103651 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49066 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100574 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18456 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26614 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/75007 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32163 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101534 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/14000 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88987 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55365 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13780 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6952 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48500 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83736 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7030 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106024 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25620 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18659 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83978 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25997 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85188 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83463 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21085 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28108 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5778 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19298 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25578 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25396 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28716 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26971 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Tools/WebKitTestRunner/mac/PlatformWebViewMac.mm:
```diff
@@ -74,16 +74,15 @@ - (void)_setWindowResolution:(CGFloat)resolution displayIfChanged:(BOOL)displayI
-    CGFloat yOffset = m_options.shouldShowWindow() ? 100 : (firstScreen ? ([firstScreen frame].size.height - rect.size.height + 10000) : -10000);
-    NSRect windowRect = NSOffsetRect(rect, xOffset, yOffset);
+    NSScreen *firstScreen = [[NSScreen screens] firstObject];
+    RELEASE_ASSERT_WITH_MESSAGE(firstScreen, "No screens found, likely due to a WindowServer crash. This WKTR configuration is not supported.");
```
How do we know it's "likely"?  Maybe "possibly" instead?

zakariaridouh replied:
Yup, a more correct statement is that the process cant connect to the WindowServer for any variety of reasons (sandboxing, permissions etc) and not just due to a crash.

ddkilzer reviewed (approved):
r=me, but let's wait for more feedback before landing it

zakariaridouh committed:
[WKTR] Crash WKTR in setup if no screens returned from AppKit
https://bugs.webkit.org/show_bug.cgi?id=291122
rdar://148636042

Reviewed by David Kilzer and Alexey Proskuryakov.

We've come across some situations where the `NSScreen screens` API
doesn't return any valid screens for us to use in WKTR (for example
post WindowServer crash, or sandboxing where the app can't connect to WS),
and we should not support this configuration.

Add a release assert with message to crash the process.

* Tools/WebKitTestRunner/mac/PlatformWebViewMac.mm:
(WTR::PlatformWebView::PlatformWebView):

Canonical link: https://commits.webkit.org/293315@main
